controller/two_lane_queue.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"k8s.io/client-go/util/workqueue"
+)
+
+// item wraps workqueue.Interface.Get() return
+type item struct {
+	i  interface{}
+	sd bool
+}
+
+// twoLaneQueue is a rate limited queue that wraps around two queues
+// -- fast queue, whose contents are processed with priority.
+// -- slow queue, whose contents are processed if fast queue has no items.
+// All the default methods operate on the fast queue, unless noted otherwise.
+type twoLaneQueue struct {
+	workqueue.RateLimitingInterface
+	slowLane workqueue.RateLimitingInterface
+	// consumerQueue is necessary to ensure that we're not reconciling
+	// the same object at the exact same time (e.g. if it had been enqueued
+	// in both fast and slow and is the only object there).
+	consumerQueue workqueue.Interface
+
+	name string
+
+	fastChan chan item
+	slowChan chan item
+}
+
+// Creates a new twoLaneQueue.
+func newTwoLaneWorkQueue(name string) *twoLaneQueue {
+	rl := workqueue.DefaultControllerRateLimiter()
+	tlq := &twoLaneQueue{
+		RateLimitingInterface: workqueue.NewNamedRateLimitingQueue(
+			rl,
+			name+"-fast",
+		),
+		slowLane: workqueue.NewNamedRateLimitingQueue(
+			rl,
+			name+"-slow",
+		),
+		consumerQueue: workqueue.NewNamed(name + "-consumer"),
+		name:          name,
+		fastChan:      make(chan item),
+		slowChan:      make(chan item),
+	}
+	tlq.runConsumer()
+	tlq.runProducers()
+	return tlq
+}
+
+func process(q workqueue.Interface, ch chan item) {
+	go func(q workqueue.Interface) {
+		for {
+			i, d := q.Get()
+			// If the queue is empty and we're shutting down — stop the loop.
+			if d {
+				break
+			}
+			ch <- item{i: i, sd: d}
+		}
+		// Sender closes the channel
+		close(ch)
+	}(q)
+}
+
+func (tlq *twoLaneQueue) runProducers() {
+	process(tlq.RateLimitingInterface, tlq.fastChan)
+	process(tlq.slowLane, tlq.slowChan)
+}
```

Comment:
```suggestion
func process(q workqueue.Interface, ch chan item) {
	// Sender closes the channel
	defer close(ch)
	for {
		i, d := q.Get()
		// If the queue is empty and we're shutting down — stop the loop.
		if d {
			return
		}
		ch <- item{i: i, sd: d}
	}
}

func (tlq *twoLaneQueue) runProducers() {
	go process(tlq.RateLimitingInterface, tlq.fastChan)
	go process(tlq.slowLane, tlq.slowChan)
}
```

Reply:
IMO when this processor has sent the item from `Get()` along the channel, that workqueue is `Done()` with that item.

Reply:
Sure that makes sense.